operator/pkg/ciliumendpointbatch/endpointbatch.go

```diff
@@ -0,0 +1,300 @@
+	log.Info("Remove stale CEP entries in CEB")
+
+	// Get all ceps from local datastore
+	staleCeps := c.Manager.getAllCepNames()
+
+	// Remove stale CEP entries present in CEB
+	for _, cepName := range staleCeps {
+		if _, exists, err := c.ciliumEndpointStore.GetByKey(cepName); err == nil && exists || err != nil {
+			continue
+		}
+		log.WithFields(logrus.Fields{
+			"cep-name": cepName,
+		}).Debug("Removing stale CEP entry.")
+		c.Manager.RemoveCepFromCache(cepName)
+	}
+	return
+}
+
+// Sync all cebs from cebStore to manager cache.
+// Note: CEBs are synced locally before CEB controller running and this is required.
+func syncCebsInLocalCache(cebStore cache.Store, manager cebManager) {
+
+	for _, obj := range cebStore.List() {
+		ceb := obj.(*v2alpha1.CiliumEndpointBatch)
+		// If CEB is already cached locally, do nothing.
+		if _, err := manager.getCebFromCache(ceb.GetName()); err == nil {
+			continue
+		}
+
+		// Create new CEB locally, with the given cebName
+		manager.createCeb(ceb.GetName())
+
+		// Deep copy the ceb, we got from api-server to local datastore.
+		manager.updateCebInCache(ceb, true)
+
+	}
+	log.Debug("Successfully synced all CEBs locally")
+	return
+}
+
+// worker runs a worker thread that just dequeues items, processes them, and
+// marks them done. You may run as many of these in parallel as you wish; the
+// workqueue guarantees that they will not end up processing the same ceb
+// at the same time
+func (c *CiliumEndpointBatchController) worker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+func (c *CiliumEndpointBatchController) processNextWorkItem() bool {
+	cKey, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(cKey)
+
+	err := c.syncCeb(cKey.(string))
+	c.handleErr(err, cKey)
+
+	return true
+}
+
+// TODO: Add metrics for failure and success rate of CEBs syncing with k8s-apiserver
+func (c *CiliumEndpointBatchController) handleErr(err error, key interface{}) {
+
+	if err == nil {
+		c.queue.Forget(key)
+		return
+	}
+
+	if c.queue.NumRequeues(key) < maxRetries {
+		c.queue.AddRateLimited(key)
+		return
+	}
+	// Drop the CEB from queue, we maxed out retries.
+	c.queue.Forget(key)
+}
+
+// syncCeb reconciles the queued CEB with api-server.
+func (c *CiliumEndpointBatchController) syncCeb(key string) error {
+
+	// Check the CEB exists is in cebStore i.e. in api-server copy of CEBs, if exist update or delete the CEB.
+	obj, exists, err := c.ciliumEndpointBatchStore.GetByKey(key)
+	if err == nil && exists {
+		ceb := obj.(*v2alpha1.CiliumEndpointBatch)
+		// Delete the CEB, only if CEP count is zero in local copy of CEB and api-server copy of CEB,
+		// else Update the CEB
+		localCeb, _ := c.Manager.getCebFromCache(key)
+		if len(ceb.Endpoints) == 0 && len(localCeb.Endpoints) == 0 {
+			if err := c.reconciler.reconcileCebDelete(key); err != nil {
+				return err
+			}
+		} else {
+			if err := c.reconciler.reconcileCebUpdate(key); err != nil {
```

Comment:
should we delete instead of updating if `len(localCeb.Endpoints) == 0`?

Reply:
The reason to check length of `local ceb cache` and `ceb watcher cache` is avoid all race condition issues. If we send `delete` request instead of `update` and meantime if some CEPs are regenerated with same namespace/name, we will be updating `local ceb cache` again. That means, from operator perspective that CEB is alive.

Reply:
> If we send delete request instead of update and meantime if some CEPs are regenerated with same namespace/name, we will be updating local ceb cache again. That means, from operator perspective that CEB is alive.

Isn't this true for your logic inside above `if` condition as well?

Reply:
sorry i'm not following, there were few `if` can you paste snippet of code ?

Reply:
sry, I meant line 269:
```
		if len(ceb.Endpoints) == 0 && c.Manager.getCepCountInCeb(key) == 0 {
			if err := c.reconciler.reconcileCebDelete(key); err != nil {
				return err
			}
```

Reply:
Synced offline, mark as resolved.